### ISSUE TITLE
你好，代码可能有个小bug

### ISSUE THREAD

**Issue by jj199603** [MJ-NCEPU/DiffJPEG#1]
原版的DIFFJPEG代码做量化时的操作如下：
image = image.float() / (utils.y_table * factor)
您的代码相当于让batch中的每一个样本都传入了不同的factor，这很好，但我认为您的代码中
image = image.float() / (self.y_table @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1)应该改成
image = image.float() / (self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1)
因为self.y_table.T @ self.factor相当于把batch_size得到的shape是(8, 8*batch_size)，相当于batch_size个矩阵沿着dim=1去cat，然后你对这个结果转置再reshape得到的batch_size个8*8矩阵都是原来代码得到的矩阵的转置

**Comment by jj199603** [comment]
用您的代码可以做个简单实验：
factor = quality_to_factor(quality=20, batch=3)
result1 = (self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1)
result2 = self.y_table * (5000./20/100.)原版是这样

**Comment by MJ-NCEPU** [comment]
> 用您的代码可以做个简单实验： factor = quality_to_factor(quality=20, batch=3) result1 = (self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1) result2 = self.y_table * (5000./20/100.)原版是这样

看到你的信息了，这一块的改动我当时也实验了好几次，当我改过后为了验证是否正确，就把图片按Batch输入，然后我看结果和压缩因子是一一对应的。你可以按照你的想法修改然后测试一下结果，比如batch=8或16或32。

**Comment by MJ-NCEPU** [comment]
> 用您的代码可以做个简单实验： factor = quality_to_factor(quality=20, batch=3) result1 = (self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1) result2 = self.y_table * (5000./20/100.)原版是这样
开源包Basicsr里面也有DiffJpeg的代码，可以直接copy,但batch处理方式应该不一样。

**Comment by jj199603** [comment]
感谢回复，我已经做过实验了，(self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1)和原版得出的结果是一样的，我也不太明白您说的结果和压缩因子是一一对应是什么原理

**Comment by MJ-NCEPU** [comment]
好的，我稍后再实验一下，感谢指正。



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年4月6日(周四) 中午11:27
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [MJ-NCEPU/DiffJPEG] 你好，代码可能有个小bug (Issue #1)




 
感谢回复，我已经做过实验了，(self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1)和原版得出的结果是一样的，我也不太明白您说的结果和压缩因子是一一对应是什么原理

**Comment by jj199603** [comment]
我没有完整地跑你的代码，只是看了一下(self.y_table.T @ self.factor).T.reshape(self.arg.batch_size, 8, 8).unsqueeze(1)算出来的量化table应该是原版一样的，如果不加.T是转置的

**Comment by jj199603** [comment]
这是我的完整测试代码，你可以参考一下：
import numpy as np
# PyTorch
import torch
import torch.nn as nn

y_table = np.array(
    [[16, 11, 10, 16, 24, 40, 51, 61], [12, 12, 14, 19, 26, 58, 60,
                                        55], [14, 13, 16, 24, 40, 57, 69, 56],
     [14, 17, 22, 29, 51, 87, 80, 62], [18, 22, 37, 56, 68, 109, 103,
                                        77], [24, 35, 55, 64, 81, 104, 113, 92],
     [49, 64, 78, 87, 103, 121, 120, 101], [72, 92, 95, 98, 112, 100, 103, 99]],
    dtype=np.float32).T

y_table = torch.from_numpy(y_table)

def quality_to_factor(quality, batch):
    """ Calculate factor corresponding to quality
    Input:
        quality(float): Quality for jpeg compression
    Output:
        factor(float): Compression factor
    """
    if isinstance(quality, int):
        quality = torch.Tensor(np.repeat(quality, batch))
    qua = []
    for qi in quality:
        eye = torch.eye(8)
        if qi < 50:
            qi = 5000. / qi
        else:
            qi = 200. - qi * 2
        qua.append(eye * (qi / 100.))
    return torch.cat(qua, dim=1)
    

factor = quality_to_factor(20, 3)
img = torch.rand(3, 1024, 8, 8)
result1 = img / (y_table.T @ factor).T.reshape(3, 8, 8).unsqueeze(1)
# result1 = img / (y_table @ factor).T.reshape(3, 8, 8).unsqueeze(1)
result2 = img / (y_table * (5000./20/100.))  # 原版DIFFJPEG
print(result1==result2)
print(torch.count_nonzero(result1-result2))

**Comment by MJ-NCEPU** [comment]
好的好的



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年4月6日(周四) 中午11:35
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [MJ-NCEPU/DiffJPEG] 你好，代码可能有个小bug (Issue #1)




 
这是我的完整测试代码，你可以参考一下：
 import numpy as np
 
PyTorch
 
import torch
 import torch.nn as nn
 
y_table = np.array(
 [[16, 11, 10, 16, 24, 40, 51, 61], [12, 12, 14, 19, 26, 58, 60,
 55], [14, 13, 16, 24, 40, 57, 69, 56],
 [14, 17, 22, 29, 51, 87, 80, 62], [18, 22, 37, 56, 68, 109, 103,
 77], [24, 35, 55, 64, 81, 104, 113, 92],
 [49, 64, 78, 87, 103, 121, 120, 101], [72, 92, 95, 98, 112, 100, 103, 99]],
 dtype=np.float32).T
 
y_table = torch.from_numpy(y_table)
 
def quality_to_factor(quality, batch):
 """ Calculate factor corresponding to quality
 Input:
 quality(float): Quality for jpeg compression
 Output:
 factor(float): Compression factor
 """
 if isinstance(quality, int):
 quality = torch.Tensor(np.repeat(quality, batch))
 qua = []
 for qi in quality:
 eye = torch.eye(8)
 if qi < 50:
 qi = 5000. / qi
 else:
 qi = 200. - qi * 2
 qua.append(eye * (qi / 100.))
 return torch.cat(qua, dim=1)
 
factor = quality_to_factor(20, 3)
 img = torch.rand(3, 1024, 8, 8)
 result1 = img / (y_table.T @ factor).T.reshape(3, 8, 8).unsqueeze(1)
 
result1 = img / (y_table @ factor).T.reshape(3, 8, 8).unsqueeze(1)
 
result2 = img / (y_table * (5000./20/100.))  # 原版DIFFJPEG
 print(result1==result2)
 print(torch.count_nonzero(result1-result2))

**Comment by jj199603** [comment]
其实我并不是说您的代码不对，只是说不加转置，可能算出来的那个量化table和原版是倒置的

**Comment by MJ-NCEPU** [comment]
哈哈，没事没事。



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年4月6日(周四) 中午11:36
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [MJ-NCEPU/DiffJPEG] 你好，代码可能有个小bug (Issue #1)




 
其实我并不是说您的代码不对，只是说不加转置，可能算出来的那个量化table和原版是倒置的

**Comment by jj199603** [comment]
期待您的验证和回复~
